File: src/main/scala/mffs/base/TileFrequency.java
package mffs.base;

import resonant.api.mffs.IBiometricIdentifierLink;
import resonant.api.mffs.card.ICoordLink;
import resonant.api.mffs.fortron.FrequencyGrid;
import resonant.api.mffs.security.IBiometricIdentifier;
import com.google.common.io.ByteArrayDataInput;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.tileentity.TileEntity;
import resonant.api.blocks.IBlockFrequency;
import universalelectricity.core.transform.vector.Vector3;

import java.io.IOException;
import java.util.HashSet;
import java.util.Set;

public abstract class TileFrequency extends TileMFFSInventory implements IBlockFrequency, IBiometricIdentifierLink
{
	private int frequency;

	@Override
	public void initiate()
	{
		FrequencyGrid.instance().register(this);
		super.initiate();
	}

	@Override
	public void invalidate()
	{
		FrequencyGrid.instance().unregister(this);
		super.invalidate();
	}

	@Override
	public void onReceivePacket(int packetID, ByteArrayDataInput dataStream) throws IOException
	{
		super.onReceivePacket(packetID, dataStream);

		if (packetID == TilePacketType.FREQUENCY.ordinal())
		{
			this.setFrequency(dataStream.readInt());
		}
	}

	@Override
	public void readFromNBT(NBTTagCompound nbt)
	{
		super.readFromNBT(nbt);
		this.setFrequency(nbt.getInteger("frequency"));
	}

	@Override
	public void writeToNBT(NBTTagCompound nbt)
	{
		super.writeToNBT(nbt);
		nbt.setInteger("frequency", this.getFrequency());
	}

	@Override
	public int getFrequency()
	{
		return this.frequency;
	}

	@Override
	public void setFrequency(int frequency)
	{
		this.frequency = frequency;
	}

	/**
	 * Gets the first linked security station, based on the card slots and frequency.
	 *
	 * @return
	 */
	@Override
	public IBiometricIdentifier getBiometricIdentifier()
	{
		/**
		 * Try to find in the cards first.
		 */
		if (this.getBiometricIdentifiers().size() > 0)
		{
			return (IBiometricIdentifier) this.getBiometricIdentifiers().toArray()[0];
		}

		return null;
	}

	@Override
	public Set<IBiometricIdentifier> getBiometricIdentifiers()
	{
		Set<IBiometricIdentifier> list = new HashSet<IBiometricIdentifier>();

		/**
		 * Try to find in the cards first.
		 */
		for (ItemStack itemStack : this.getCards())
		{
			if (itemStack != null && itemStack.getItem() instanceof ICoordLink)
			{
				Vector3 linkedPosition = ((ICoordLink) itemStack.getItem()).getLink(itemStack);

				if (linkedPosition != null)
				{
					TileEntity tileEntity = linkedPosition.getTileEntity(this.worldObj);

					if (linkedPosition != null && tileEntity instanceof IBiometricIdentifier)
					{
						list.add((IBiometricIdentifier) tileEntity);
					}
				}
			}
		}

		for (IBlockFrequency tileEntity : FrequencyGrid.instance().get(this.getFrequency()))
		{
			if (tileEntity instanceof IBiometricIdentifier)
			{
				list.add((IBiometricIdentifier) tileEntity);
			}
		}

		return list;
	}
}

File: src/main/scala/mffs/base/TileFrequency.scala
package mffs.base

import java.util.{Set => JSet}

import com.resonant.core.access.Permission
import mffs.api.Frequency
import mffs.api.card.ICoordLink
import mffs.item.card.ItemCardFrequency
import mffs.security.{BlockBiometric, MFFSPermissions}
import mffs.{GraphFrequency, Reference}
import nova.core.entity.Entity
import nova.core.item.Item
import nova.core.util.transform.Vector3d

abstract class TileFrequency extends TileMFFSInventory with Frequency {
	val frequencySlot = 0

	override def load() {
		super.load()
		GraphFrequency.instance.add(this)
	}

	override def unload() {
		super.unload()
		GraphFrequency.instance.remove(this)
	}

	def hasPermission(profile: GameProfile, permissions: Permission*): Boolean = permissions.forall(hasPermission(profile, _))

	def hasPermission(profile: GameProfile, permission: Permission): Boolean = !isActive || getBiometricIdentifiers.forall(_.hasPermission(profile, permission))

	def getBiometricIdentifiers: Set[BlockBiometric] = {
		val cardLinks = getConnectionCards.view
			.filter(item => item != null && item.isInstanceOf[ICoordLink])
			.map(item => item.asInstanceOf[ICoordLink].getLink())
			.filter(_ != null)
			.filter(_.isInstanceOf[BlockBiometric])
			.map(_.asInstanceOf[BlockBiometric])
			.force
			.toSet

		val frequencyLinks = GraphFrequency.instance.get(getFrequency).filter(_.isInstanceOf[BlockBiometric]).toSet[BlockBiometric]

		return frequencyLinks ++ cardLinks
	}

	override def getFrequency: Int = {
		val frequencyCard = getFrequencyCard

		if (frequencyCard != null) {
			return frequencyCard.asInstanceOf[ItemCardFrequency].getFrequency(frequencyCard)
		}

		return 0
	}

	override def setFrequency(frequency: Int) {

	}

	def getFrequencyCard: ItemCardFrequency = {
		val stack = inventory.get(frequencySlot)

		if (stack != null && stack.isInstanceOf[ItemCardFrequency]) {
			return stack.asInstanceOf[ItemCardFrequency]
		}

		return null
	}

	/**
	 * Gets a set of cards that define connections.
	 */
	def getConnectionCards: Set[Item] = Set(inventory.get(0).orElseGet(null))

	/**
	 * Gets the first linked biometric identifier, based on the card slots and frequency.
	 */
	def getBiometricIdentifier: BlockBiometric = if (getBiometricIdentifiers.size > 0) getBiometricIdentifiers.head else null

	override def onRightClick(entity: Entity, side: Int, hit: Vector3d): Boolean = {
		if (!hasPermission(player.getGameProfile, MFFSPermissions.configure)) {
			player.addChatMessage(new ChatComponentText("[" + Reference.name + "]" + " Access denied!"))
			return false
		}

		return super.onRightClick(entity, side, hit)
	}
}